src/nets/me2e_asr_attctc.py

```diff
@@ -0,0 +1,204 @@
+
+    def __init__(self, cmvn_stats):
+        super(CMVN, self).__init__()
+
+        self.cmvn_mean = cmvn_stats[0]
+        self.cmvn_std = cmvn_stats[1]
+
+    def __call__(self, feats, xp):
+        cmvn_mean = chainer.Variable(xp.array(self.cmvn_mean, dtype=np.float32))
+        cmvn_std = chainer.Variable(xp.array(self.cmvn_std, dtype=np.float32))
+
+        cmvn_feats = []
+        for feat in feats:
+            feat -= F.broadcast_to(cmvn_mean, shape=feat.shape)
+            feat /= F.broadcast_to(cmvn_std, shape=feat.shape)
+            cmvn_feats.append(feat)
+
+        return cmvn_feats
+
+
+class ME2E(chainer.Chain):
+    def __init__(self, enhan, asr, melmat, cmvn_stats):
+        super(ME2E, self).__init__()
+
+        with self.init_scope():
+            # enhan
+            self.enhan = enhan
+            # asr
+            self.asr = asr
+
+        self.logmel = LogMel(melmat)
+        self.cmvn = CMVN(cmvn_stats)
+
+    # x[i]: ('utt_id', {'ilen':'xxx',...}})
+    def __call__(self, data):
+        '''ME2E forward
+
+        :param data:
+        :return:
+        '''
+        mode = data[0][1]['mode']
+        logging.info('mode: ' + mode)
+
+        if mode == 'noisy':
+            # utt list of frame x dim
+            xs_real = [i[1]['feat']['real'] for i in data]
+            xs_imag = [i[1]['feat']['imag'] for i in data]
+            ilens = self.xp.array([utt.shape[0] for utt in xs_real], dtype=np.int32)
+            hs_real = [chainer.Variable(self.xp.array(utt, dtype=np.float32)) for utt in xs_real]
+            hs_imag = [chainer.Variable(self.xp.array(utt, dtype=np.float32)) for utt in xs_imag]
+
+            hs = {}
+            hs['real'] = hs_real
+            hs['imag'] = hs_imag
+        elif mode == 'enhan':
+            # utt list of channel list of frame x dim
+            xs_real = [i[1]['feat']['real'] for i in data]
+            xs_imag = [i[1]['feat']['imag'] for i in data]
+            ilens = self.xp.array([utt[0].shape[0] for utt in xs_real], dtype=np.int32)
+            hs_real = [[chainer.Variable(self.xp.array(ch, dtype=np.float32))
+                        for ch in utt] for utt in xs_real]
+            hs_imag = [[chainer.Variable(self.xp.array(ch, dtype=np.float32))
+                        for ch in utt] for utt in xs_imag]
+
+            hs = {}
+            hs['real'] = hs_real
+            hs['imag'] = hs_imag
+
+            # 1. beamformer
+            hs = self.enhan(hs)
+        else:
+            logging.error(
+                "Error: need to specify an appropriate training mode")
+            sys.exit()
+
+        # utt list of olen
+        ys = [self.xp.array(
+            list(map(int, i[1]['tokenid'].split())), dtype=np.int32) for i in data]
+        ys = [chainer.Variable(y) for y in ys]
+
+        # 2. feature extractor
+        hs = self.logmel(hs, self.xp)
+        hs = self.cmvn(hs, self.xp)
+
+        # 3. encoder
+        hs, ilens = self.asr.enc(hs, ilens)
```

Comment:
if following lines are completely same in `E2E.__call__`, you can separate those lines in a separated method like
``` python
    def __call__(self, data):
        '''E2E forward
        :param data:
        :return:
        '''
        # utt list of frame x dim
        xs = [i[1]['feat'] for i in data]
        # remove 0-output-length utterances
        tids = [d[1]['tokenid'].split() for d in data]
        filtered_index = list(filter(lambda i: len(tids[i]) > 0, range(len(xs))))
        if len(filtered_index) != len(xs):
            logging.warning('Target sequences include empty tokenid (batch %d -> %d).' % (
                len(xs), len(filtered_index)))
        xs = [xs[i] for i in filtered_index]
        # utt list of olen
        ys = [self.xp.array(np.fromiter(map(int, tids[i]), dtype=np.int32))
              for i in filtered_index]
        ys = [chainer.Variable(y) for y in ys]

        # subsample frame
        xs = [xx[::self.subsample[0], :] for xx in xs]
        ilens = self.xp.array([xx.shape[0] for xx in xs], dtype=np.int32)
        hs = [chainer.Variable(self.xp.array(xx, dtype=np.float32))
              for xx in xs]
        return self.loss(hs, ilens, ys)

    def loss(self, hs, ilens, ys)
        # 1. encoder
        hs, ilens = self.enc(hs, ilens)

        # 3. CTC loss
        loss_ctc = self.ctc(hs, ys)

        # 4. attention loss
        loss_att, acc, att_w = self.dec(hs, ys)
        return loss_ctc, loss_att, acc
```

Reply:
Basecally, the differences between E2E and ME2E are data preparation and beamforming computation parts.
After those processes, the code is completely same.
However, because data preparation part is different, we cannot call E2E's __call__ function.
If E2E has just network computing function like the above loss, we can just import and use it...

Reply:
OK, I know. If you are not comfortable with this change, you can keep it.

Reply:
Sorry, did you propose the modification of the original E2E function (__call__ and loss)?
If you can modify so, I can just call E2E.loss in the ME2E's __call__ function.

Reply:
Yes that is right. I want you to modify `E2E` as you know.